linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,330 @@
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_timer_cancel, TIMER_INSTANCE_HANDLE, timer);
+```
+
+`threadpool_timer_cancel` shall stops the timer. Afterward, the timer may be resumed with a new time by calling `threadpool_timer_restart` or cleaned up by calling `threadpool_timer_destroy`.
+
+**SRS_THREADPOOL_LINUX_11_050: [** If `timer` is `NULL`, `threadpool_timer_cancel` shall fail and return. **]**
+
+**SRS_THREADPOOL_LINUX_11_051: [** `threadpool_timer_cancel` shall call `timer_settime` with 0 for `flags` and `NULL` for `old_value` and `{0}` for `new_value` to cancel the ongoing timers. **]**
+
+### threadpool_timer_destroy
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_timer_destroy, TIMER_INSTANCE_HANDLE, timer);
+```
+
+`threadpool_timer_destroy` stops the timer started by `threadpool_timer_start` and cleans up its resources.
+
+**SRS_THREADPOOL_LINUX_11_052: [** If `timer` is `NULL`, `threadpool_timer_destroy` shall fail and return. **]**
+
+**SRS_THREADPOOL_LINUX_11_053: [** `threadpool_timer_cancel` shall call `timer_delete`  to destroy the ongoing timers. **]**
+
+**SRS_THREADPOOL_LINUX_11_054: [** `threadpool_timer_destroy` shall free all resources in `timer`. **]**
+
+### threadpool_work_func
+
+```C
+static int threadpool_work_func(void* param);
+```
+
+`threadpool_work_func` shall obtain a work function.
+
+**SRS_THREADPOOL_LINUX_11_055: [** If `param` is `NULL`, `threadpool_work_func` shall return. **]**
+
+**SRS_THREADPOOL_LINUX_11_056: [** `threadpool_work_func` shall get the real time by calling `clock_gettime`. **]**
+
+**SRS_THREADPOOL_LINUX_11_080: [** If `clock_gettime` fails, `threadpool_work_func` shall return. **]**
+
+**SRS_THREADPOOL_LINUX_11_081: [** If `sem_timedwait` fails, `threadpool_work_func` shall timeout and return. **]**
+
+**SRS_THREADPOOL_LINUX_11_057: [** `threadpool_work_func` shall decrement the threadpool semaphore with a time limit for 2 seconds. **]**
+
+**SRS_THREADPOOL_LINUX_11_058: [** `threadpool_work_func` shall acquire the shared SRW lock by calling `srw_lock_acquire_shared`. **]**
+
+**SRS_THREADPOOL_LINUX_11_059: [** `threadpool_work_func` shall get the current task array size and next waiting task consume index. **]**
+
+**SRS_THREADPOOL_LINUX_11_060: [** If consume index has task state `TASK_WAITING`, `threadpool_work_func` shall set the task state to `TASK_WORKING`. **]**
+
+**SRS_THREADPOOL_LINUX_11_061: [** `threadpool_work_func` shall initialize `task_func` and `task_param` and then set the task state to `TASK_NOT_USED`. **]**
+
+**SRS_THREADPOOL_LINUX_11_062: [** `threadpool_work_func` shall release the shared SRW lock by calling `srw_lock_release_shared`. **]**
+
+**SRS_THREADPOOL_LINUX_11_063: [** If `task_param` is not `NULL`, `threadpool_work_func` shall execute it with parameter `task_param`. **]**
+
+### reallocate_threadpool_array
+
+```C
+static int reallocate_threadpool_array(THREADPOOL* threadpool);
+```
+
+`reallocate_threadpool_array` resize the task array size by 2.
```

Comment:
>e by 2. [](http://example.com/codeflow?start=55&length=7)

nit: "by doubling" or "multiplies the task array size by 2" or something. This sounds like it "adds 2".

Reply:
done